drafts/are-you-getting-the-test-pyramid-wrong.md

```diff
@@ -0,0 +1,194 @@
+
+        when(restTemplate.exchange(ArgumentMatchers.<?>any(), ArgumentMatchers.<ParameterizedTypeReference<?>>any())).thenAnswer(call -> {
+            var entity = call.getArgument(0, RequestEntity.class);
+            // Check the report ID
+            ...
+            // Check the requested content type
+            ...
+            return new ResponseEntity<>(reportResponse, HttpStatus.OK);
+        });
+
+        var response = underTest.downloadReport(reportId);
+
+        assertThat(response.isRight()).isTrue();
+        assertThat(response.get().contentType()).isEqualTo(MediaType.APPLICATION_XML);
+        assertThat(new String(response.get().contents()).isEqualTo("...");
+    }
+}
+```
+
+The unit test for the `ReportStorageService` might look like this:
+
+```java
+@ExtendWith(MockitoExtension.class)
+class ReportStorageServiceTest {
+    @Mock
+    private ReportObjectToReportEntityMapper mapper;
+
+    @Mock
+    private ReportEntityRepository repository;
+
+    private ReportStorageService underTest = new ReportStorageServiceImpl(mapper, repository);
+
+    ...
+
+    @Test
+    void testStoreReport() {
+        var captor = ArgumentCaptor.forClass(ReportEntity.class);
+        var report = ...
+        var entity = ...
+        when(mapper.map(any())).thenReturn(entity);
+
+        underTest.storeReport(mapper.map(report));
+
+        verify(repository, times(1)).save(captor.capture());
+
+        var capturedEntity = captor.getValue();
+
+        assertThat(capturedEntity.getId()).isNotNull();
+        // Assert all entity fields
+        ...
+    }
+}
+```
+
+## Why is this bad?
+
+We have mocked everything except our "unit", and we're asserting almost everything possible. The tests are passing, and the application code is working fine. So why is this bad?
+
+Let's take a closer look at the `ReportDownloaderTest`. First, the `RestTemplate.exchange()` method has many overloaded versions, but we're only mocking one of them. This works because we rely on our knowledge of the implementation details. We know that we called a specific `RestTemplate.exchange()` method. But this is fragile and it would break easily. Second, the test requires us to mock the `RestTemplate.getUriTemplateHandler()` method, which is used by the Spring Framework internally and not directly by the application code. So we're mocking something that doesn't belong to us. Third, by mocking `RestTemplate` methods, we're essentially writing more test code than application code. This is a waste of time and effort, and it makes the tests more difficult to maintain.
+
+In the `ReportStorageServiceTest`, we've mocked our repository and verified that its `save()` method is called. This is not ideal because we're testing the interactions, not the behaviours. In other words, we're testing that the code calls the `save()` method, but we're not testing that the `save()` method actually persists in the data in the database. This can lead to two problems:
```

Comment:
```suggestion
In the `ReportStorageServiceTest`, we've mocked our repository and verified that its `save()` method is called. This is not ideal because we're testing the interactions, not the behaviors. In other words, we're testing that the code calls the `save()` method, but we're not testing that the `save()` method actually persists in the data in the database. This can lead to two problems:
```